core-bundle/src/Resources/contao/library/Contao/TemplateInheritance.php

```diff
@@ -160,8 +161,13 @@ public function inherit()
 			}
 		}
 
-		// Replace insert tags
-		$strBuffer = System::getContainer()->get(InsertTagParser::class)->replace($strBuffer);
+		// Replace insert tags, but not for the back end
+		$request = $container->get('request_stack')->getCurrentRequest();
+
+		if (null === $request || !$container->get('contao.routing.scope_matcher')->isBackendRequest($request))
```

Comment:
Shouldn't we use `isFrontendRequest()` instead? Because we only want insert tags to be replaced in the front end and not in any other scope (even though there is none yet).

Reply:
I am not sure. For example, if a template is parsed on the console and the insert tags are not automatically replaced, they need to be replaced manually, if you need them replaced. However, at that point you lose the granularity of replacing insert tags per template again, rather than replacing them on the finished string.

imho with the new concept (replacing in `TemplateInheritance::inherit`) the insert tags should always be replaced, except for the back end.

But then again, the new concept might actually be a BC break since some applications might not want insert tags to be replaced automatically when using `$template->parse()` and thus may be we need to replace insert tags _only_ for the Contao frontend scope indeed. But then may be insert tags should be replaced somewhere else entirely anyway?

@contao/developers wdyt?

Reply:
> Because we only want insert tags to be replaced in the front end and not in any other scope (even though there is none yet).

You could also argue, that you only don't want them to be replaced in the backend.

May there is at some point a scope `API` or the like where it also should be replaced I think.

Reply:
IMO the check for the backend scope is correct. Can we additionally add a parameter to `$template->parse()` that controls if insert tags should be replaced?

- `$template->parse()` => replace insert-tags if not in backend scope
- `$template->parse(true)` => replace insert-tags always
- `$template->parse(false)` => do not replace insert-tags

Reply:
Do you want to add the parameter in this PR or in a separate one?

Reply:
I'll add it to this PR :)

Reply:
> Can we additionally add a parameter to $template->parse() that controls if insert tags should be replaced?

I'm not sure about this one. This would suggest that insert tags are replaced with `true` even though they wouldn't if I use a Twig template. If we add a parameter I'd go for somthing like `$neverReplace = false` that can be set to `true`, but I'd just omit it tbh.

Reply:
Done. Not sure if I like though 😬

Reply:
@fritzmg Please revert the last commit. @m-vo has explained why the parameter is not a good idea.

Reply:
Right, haven't seen that comment. But still, since insert tags are now _always_ replaced now (except for the back end), which was previously not the case, shouldn't we allow to disable replacing insert tags _somehow_? May be via the Insert Tag service?

```php
$this->insertTags->pushDisable();
$template->parse();
$this->insertTags->popDisable();
```

Reply:
They were always replaced ever since, just at higher granularity level. And they are never replaced in Twig. 

The only thing I could imagine is someone using templates that aren't output in the frontend and which includes patterns used that look like insert tags `{{foo}}` that now get removed.

Reply:
> They were always replaced ever since, just at higher granularity level. And they are never replaced in Twig.

No, that's not true. They were never replaced with `$template->parse()`, only with `$template->output()`.

Reply:
Yes, I meant 'if output in the front end' which is where templates are probably used 99% of the time. :smile:

Reply:
Just to reiterate: this is where insert tags are replaced for the front end in Contao 4.9:

https://github.com/contao/contao/blob/0406c1b94e4780b021c44bf646dac7392ff5ce36/core-bundle/src/Resources/contao/classes/FrontendTemplate.php#L127

Executing

```php
$template = new FrontendTemplate('foobar');
$buffer = $template->parse();
```

will result in insert tags being present and unreplaced in `$buffer`. `FrontendTemplate` will replace insert tags _only_ in `FrontendTemplate::compile` and `FrontendTemplate::compile` will only be executed when using `FrontendTemplate::output` or `FrontendTemplate::getResponse`.

Reply:
And @andyschempp 's PR https://github.com/contao/contao/pull/3601 introduced a new `FragmentTemplate` class, which overrides the `compile` method, so that it will only execute `$this->parse()` without replacing any insert tags (or doing any other tasks that are actually only relevant to the `fe_page` tepmlate.

Reply:
We should go with something like `parse($skipReplaceInsertTags = false)` then I think. Twig templates however should ignore this parameter completely I think.

Reply:
So may be we actually need to revert https://github.com/contao/contao/pull/3622/ in order to keep backwards compatibility. I do think it is a problem that `$template->parse()` now replaces insert tags without any control over it.

Reply:
Should we discuss this in our call later this day? Might be easier :relaxed:.

Reply:
As discussed in the Contao call, we want to reverse the logic and only replace insert tags in the front end. We also want to add the argument to the `parse()` method.

Reply:
> As discussed in the Contao call, we want to reverse the logic and only replace insert tags in the front end. We also want to add an `$preserveInsertTagsInLegacyTemplates` argument to the `parse()` method.

Done in f0dd1ca